### ISSUE TITLE
eslint-disable-next-line works in shebang comments

### ISSUE THREAD

**Issue by not-an-aardvark** [eslint/eslint#9238]
**Tell us about your environment**

* **ESLint Version:** 4.6.1
* **Node Version:** 8.4.0
* **npm Version:** 5.3.0

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {
    rules: {
        no-undef: "error"
    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue.**

```js
#! eslint-disable-next-line
foo;
```

**What did you expect to happen?**

I expected an error to be reported, because [`eslint-disable-next-line`](https://eslint.org/docs/user-guide/configuring#disabling-rules-with-inline-comments) is only supposed to work when using line comments.

**What actually happened? Please include the actual, raw output from ESLint.**

No error was reported, because ESLint considered the shebang to be a valid `eslint-disable-next-line` comment.

It looks like the issue is occurring as part of the inline comment processing [here](https://github.com/eslint/eslint/blob/3c41a053026b502bc8ec3890e526f45129c84ebd/lib/linter.js#L339-L399). In a previous version of ESLint, `ast.comments` did not contain shebangs, so it was safe for that logic to assume that all comments in the array had a type of either `Block` or `Line`. However, `ast.comments` now includes shebang comments (they have type `Shebang`), so this is no longer a safe assumption.

There are a few different ways that this could be fixed:

* Filter out shebangs before doing any processing on them.
* Make the logic explicitly check that a comment has type `Line` before assuming that it's a line comment.

**Comment by i-ron-y** [comment]
I had the logic explicitly check that the comment has type "Line" before checking for a eslint-disable-next-line. Now

    #! eslint-disable-next-line
    foo;

is giving an error, which is what we want. Perfect, that fixed this issue.

Unfortunately, I discovered another issue with shebangs. The following code should give three errors, for lines (1), (2), and (3).

    "foo;",                                          (1)
    "#! eslint-disable-next-line",
    "foo;",                                          (2)
    "foo;"                                           (3)

On the contrary, it results in an error only for the line (2), with lines (1) and (3) not giving any errors.

Just to make sure this isn't an issue that I have introduced with my fix, I tried this test out on the original, unmodified eslint-disable-next-line code, and this bug is also present: with the unmodified eslint-disable-next-line code, this test gives no errors.

It feels like Shebangs aren't being processed properly... ?

**Comment by not-an-aardvark** [comment]
Thanks for working on this!

> It feels like Shebangs aren't being processed properly... ?

I think shebangs are only allowed at the start of a file -- it's considered a syntax error if they appear anywhere else.

**Comment by i-ron-y** [comment]
Hmm, I wonder why it's inadvertently disabling all errors around it (except for the error on the next line, with my fix)... I'll keep at it, see what comes up!

**Comment by not-an-aardvark** [comment]
In general, no rules get run when a parsing error occurs, because no AST could be created. So if a file has invalid syntax, the parsing error is always the only error for that file.

**Comment by i-ron-y** [comment]
Ahh, that makes sense - thank you!
